### ISSUE TITLE
direct-lvm with xfs causes Docker to hang when disk is full

### ISSUE THREAD

**Issue by samuelkarp** [moby/moby#20707]
Output of `docker version`:

```
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5/1.9.1
 Built:        
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5/1.9.1
 Built:        
 OS/Arch:      linux/amd64

```

Output of `docker info` (on a fresh host; an affected host does not respond to `docker info`):

```
Containers: 1
Images: 6
Server Version: 1.9.1
Storage Driver: devicemapper
 Pool Name: docker-docker--pool
 Pool Blocksize: 524.3 kB
 Base Device Size: 107.4 GB
 Backing Filesystem: xfs
 Data file: 
 Metadata file: 
 Data Space Used: 100.7 MB
 Data Space Total: 9.437 GB
 Data Space Available: 9.337 GB
 Metadata Space Used: 77.82 kB
 Metadata Space Total: 25.17 MB
 Metadata Space Available: 25.09 MB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.93-RHEL7 (2015-01-28)
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.1.17-22.30.amzn1.x86_64
Operating System: Amazon Linux AMI 2015.09
CPUs: 1
Total Memory: 995.6 MiB
Name: ip-172-31-34-29
ID: GO67:QAUR:ZN7D:LBC6:M366:GT2C:BQTV:F3UE:IMGD:YFNF:T37L:OQWC

```

Provide additional environment details (AWS, VirtualBox, physical, etc.): Amazon Linux AMI 2015.09.2 configured with a second EBS volume at /dev/xvdb and configured with direct LVM (using devicemapper) by `docker-storage-setup`.  By default, `docker-storage-setup` configures the thin pool originally sized at 40% of the available disk with an auto-extend policy, however the same behavior is exhibited when auto-extend is disabled and when `lv_when_full` is changed to `error`.

List the steps to reproduce the issue:
1. Configure Docker as described above (direct LVM with devicemapper using xfs)
2. Run containers to progressively consume disk (I seemed to have reasonable results using `docker run ubuntu dd if=/dev/urandom of=sample.txt bs=64M count=1 iflag=fullblock`) until the disk is full (alternately, run my [repro script](https://gist.github.com/samuelkarp/6c8b5ebc0a9cc34006c8))
3. Try to do something with Docker (any of `docker ps`, `docker info`, `docker run`, `docker stop`, `docker rm`, `docker images`, `docker rmi`)

Describe the results you received:
Docker commands hang indefinitely.  dmesg fills with things like the following:

```
[ 3029.535199] device-mapper: thin: 253:2: reached low water mark for data device: sending event.
[ 3029.589706] device-mapper: thin: 253:2: switching pool to out-of-data-space mode
[ 3037.584626] vethac84d34: renamed from eth0
[ 3037.588339] docker0: port 2(vethd9e484d) entered disabled state
[ 3037.622172] docker0: port 2(vethd9e484d) entered disabled state
[ 3037.626293] device vethd9e484d left promiscuous mode
[ 3037.629420] docker0: port 2(vethd9e484d) entered disabled state
[ 3089.592091] device-mapper: thin: 253:2: switching pool to read-only mode
[ 3089.598147] Buffer I/O error on dev dm-4, logical block 144077, lost async page write
[ 3089.603304] Buffer I/O error on dev dm-4, logical block 144078, lost async page write
[ 3089.608669] Buffer I/O error on dev dm-4, logical block 144079, lost async page write
[ 3089.614565] Buffer I/O error on dev dm-4, logical block 144080, lost async page write
[ 3089.620112] Buffer I/O error on dev dm-4, logical block 144081, lost async page write
[ 3089.624919] Buffer I/O error on dev dm-4, logical block 144082, lost async page write
[ 3089.633141] Buffer I/O error on dev dm-4, logical block 144083, lost async page write
[ 3089.638389] Buffer I/O error on dev dm-4, logical block 144084, lost async page write
[ 3089.645177] Buffer I/O error on dev dm-4, logical block 144085, lost async page write
[ 3089.650294] Buffer I/O error on dev dm-4, logical block 144086, lost async page write
[ 3089.664084] XFS (dm-4): metadata I/O error: block 0x2c230 ("xfs_buf_iodone_callbacks") error 5 numblks 16
[ 3089.684777] XFS (dm-5): metadata I/O error: block 0x2c230 ("xfs_buf_iodone_callbacks") error 5 numblks 16
[ 3089.695953] XFS (dm-4): Unmounting Filesystem
[ 3089.703478] XFS (dm-4): metadata I/O error: block 0x2c230 ("xfs_buf_iodone_callbacks") error 5 numblks 16
[ 3089.710627] XFS (dm-5): metadata I/O error: block 0x2c230 ("xfs_buf_iodone_callbacks") error 5 numblks 16
...
[ 3150.336222] XFS (dm-6): metadata I/O error: block 0x2c230 ("xfs_buf_iodone_callbacks") error 5 numblks 16
[ 3240.180102] INFO: task docker:3312 blocked for more than 120 seconds.
[ 3240.184544]       Tainted: G            E   4.1.17-22.30.amzn1.x86_64 #1
[ 3240.189691] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[ 3240.195639] docker          D ffff880037d03d48     0  3312      1 0x00000000
[ 3240.200654]  ffff880037d03d48 ffff88003e321980 ffff88003745b300 0000000000000246
[ 3240.206939]  ffff880037d04000 ffff88003d6aabe0 ffff880037e6cdc0 ffff880037e6cde8
[ 3240.212569]  ffff880037e6cd90 ffff880037d03d68 ffffffff814df3e7 ffff88003d6aabe0
[ 3240.218299] Call Trace:
[ 3240.219963]  [<ffffffff814df3e7>] schedule+0x37/0x90
[ 3240.223179]  [<ffffffffa048fb79>] xfs_ail_push_all_sync+0xa9/0xe0 [xfs]
[ 3240.227811]  [<ffffffff810aafc0>] ? prepare_to_wait_event+0x110/0x110
[ 3240.233513]  [<ffffffffa047b2e9>] xfs_unmountfs+0x59/0x170 [xfs]
[ 3240.239390]  [<ffffffffa047bdab>] ? xfs_mru_cache_destroy+0x6b/0x90 [xfs]
[ 3240.247251]  [<ffffffffa047dd36>] xfs_fs_put_super+0x36/0x90 [xfs]
[ 3240.252770]  [<ffffffff811ccb06>] generic_shutdown_super+0x76/0x100
[ 3240.256723]  [<ffffffff811ccea7>] kill_block_super+0x27/0x70
[ 3240.260913]  [<ffffffff811cd1b9>] deactivate_locked_super+0x49/0x80
[ 3240.265629]  [<ffffffff811cd7ae>] deactivate_super+0x4e/0x70
[ 3240.269797]  [<ffffffff811e9a63>] cleanup_mnt+0x43/0x90
[ 3240.274904]  [<ffffffff811e9b02>] __cleanup_mnt+0x12/0x20
[ 3240.279970]  [<ffffffff81085b97>] task_work_run+0xb7/0xf0
[ 3240.284063]  [<ffffffff81013bb1>] do_notify_resume+0x51/0x80
[ 3240.288751]  [<ffffffff814e35bc>] int_signal+0x12/0x17
```

Describe the results you expected:
Docker itself would continue to be responsive, but operations that consume space (pulling images, writing inside a container, etc) would fail.  This is the behavior observed when I switch to ext4 instead of xfs.

**Comment by samuelkarp** [comment]
There's some talk which I think is related on the [dm-devel mailing list](https://www.redhat.com/archives/dm-devel/2015-July/msg00120.html) and some [general information](https://lwn.net/Articles/592645/) on the things that are different about thin provisioning.

**Comment by thaJeztah** [comment]
I think this is a duplicate or closely related to https://github.com/docker/docker/issues/20272

**Comment by rhvgoyal** [comment]
Currently I am aware of xfs issue where if thin pool gets full, then xfs might retry infinitely a failed IO hoping it will be successful. In normal scenarios, thin pool automatically grows and IO is successful. But when underlying disk is full and there is no space to grow pool, then IO does not succeed and xfs does not know about it and keeps on retrying.

People have suggested that configure storage properly and monitor it and don't let thin pool run out of space condition to avoid this situation.

Upstream one of the solutions proposed is tunable error behavior where a user can specify what do to when error is encountered. I think those patches are not upstream yet.

So while upstream finds a solution for the issue, in short term I think users need to keep a watch on their thin pool and backing storage and make sure it does not get full. Otherwise a reboot will be required. One shall have to add more disk space, grow thin pool, check/repair thin pool and restart
docker again.

**Comment by samuelkarp** [comment]
> So while upstream finds a solution for the issue, in short term I think users need to keep a watch on their thin pool and backing storage and make sure it does not get full.

This makes me question whether xfs is an appropriate default then.  There's nothing in the [documentation](https://docs.docker.com/engine/userguide/storagedriver/device-mapper-driver/#configure-direct-lvm-mode-for-production) suggesting that running out of space in the thin pool will cause your Docker daemon to hang (and potentially cause a kworker to chew through CPU), and Docker certainly does not take preemptive action to avoid such a situation when available disk is low.

**Comment by rhvgoyal** [comment]
One reason we had switched to xfs was that ext4 was taking lot of time during mkfs.ext4 on 100GB thin volume. While xfs did not as it created some of the data dynamically.

Now we have gone back to 10G default rootfs size. So going back to ext4 will be an option.

Or some people have suggested that introduce another knob in graph driver which denies new device
creation when pool does not have much space left. Say --storage-opt dm.min_free_space=1G. Now device mapper graph driver will fail pulling images or creating new containers etc. One could start existing containers and currently running container will continue to run.

This is also not fool proof though. As if an existing container decides to write data to rootfs and fill rest of the pool, at some point of time that container might hang waiting for IO to complete.

**Comment by samuelkarp** [comment]
> One reason we had switched to xfs was that ext4 was taking lot of time during mkfs.ext4 on 100GB thin volume. While xfs did not as it created some of the data dynamically.

Were there other reasons for switching to xfs as well or was this the only reason?

> Or some people have suggested that introduce another knob in graph driver which denies new device
creation when pool does not have much space left. Say --storage-opt dm.min_free_space=1G. Now device mapper graph driver will fail pulling images or creating new containers etc. One could start existing containers and currently running container will continue to run.

Yes, that seems like it might be a good option here.

> This is also not fool proof though. As if an existing container decides to write data to rootfs and fill rest of the pool, at some point of time that container might hang waiting for IO to complete.

That makes sense, but it would at least be an improvement.

**Comment by rhvgoyal** [comment]
CC @shishir-a412ed 

shishir is looking into implementing a knob for devicemapper to stop creation of new device once free space falls below a threshold.

**Comment by snitm** [comment]
This "issue" couldn't be more improperly categorized.  By asserting "direct-lvm with xfs causes Docker to hang when disk is full" you've shown willful ignorance about _why_ XFS degenerates to busy waiting when the thin-pool runs out of space.  And the reason boils down to improper configuration and reckless use of a finite resource.

First let me say that we'd love for this not to be an issue but the fact of the matter is it is a pitfall that has _always_ been at the heart of why it is imperative for docker (and the admin) to be aware of:
1) the expected rate of change and use of free space
2) how much space is currently available

Admins cannot stuff 10 pounds into a 5 pound bag (of DM thinp on XFS) and expect things to gracefully halt with a message of "oh you wanted performance (which XFS has over ext4), etc but you didn't reason through proper resource allocation means for your use... please take a moment to think and act accordingly".  Simply put: Care must be taken to properly provision and then monitor the free space utilization over time.  XFS's existing behavior of spinning waiting for more space just happens to make this requirement much more cut throat.

Docker being aware of how much dm-thin-pool space is free and disallowing spinning up new containers, etc is a perfectly reasonable (and overdue) response to avoid the more problematic fault handling (e.g. XFS's busy waiting for more space that never gets added).  Just because ext4 can tolerate this case more gracefully doesn't mean users should be left to keep making the same mistakes over and over.  And in doing so: exposing themselves to error handling bugs that even ext4 is known to have had (at least historically).

In a properly provisioned, monitored and managed docker deployment this failure case should _never_ present itself.  _NEVER_.  If it is then the admin has failed.  And docker quite likely has failed by allowing the admin enough rope to fail.  Regardless of which underlying filesystem is used.

**Comment by samuelkarp** [comment]
> This "issue" couldn't be more improperly categorized. By asserting "direct-lvm with xfs causes Docker to hang when disk is full" you've shown willful ignorance about why XFS degenerates to busy waiting when the thin-pool runs out of space.

My apologies.  I've attempted to learn as much as I can about XFS, devicemapper, LVM, and so forth recently and my (unwilling) ignorance is clearly coming across.

> And the reason boils down to improper configuration and reckless use of a finite resource.

This sounds true.  Regardless, this is the recommended configuration in Docker's [documentation](https://docs.docker.com/engine/userguide/storagedriver/device-mapper-driver/#configure-direct-lvm-mode-for-production) on running devicemapper.

> Admins cannot stuff 10 pounds into a 5 pound bag (of DM thinp on XFS) and expect things to gracefully halt [...]

Also sensible.  I think what's missing is that xfs is now the default since Docker 1.9.1 and this behavior is not documented anywhere for people who are unfamiliar with these pitfalls.

**Comment by snitm** [comment]
Apologies if my frustration came through... I've not been able to keep up with how each project documents the importance of properly configuring, monitoring and managing the free space that is relied on for forward progress.

But for docker setups that use DM thin-provisioning they should:
1) use lvm to setup the thin-pool that docker uses for container storage
2) configure lvm such that it will resize the thin-pool, from free space in the parent VG, once the lvm.conf configured threshold is crossed
3) the parent VG's free space _must_ be monitored and replenished; otherwise the autoresize of the thin-pool that crossed the low water mark will not have more space to pull on when it needs it again

**Comment by rhvgoyal** [comment]
@samuelkarp I have written a patch which will start failing new device creation operation in thin pool once free data/metadata space in thin pool is below a specified threshold. Right now in-built default threshold is 20%. User can change it using dm.min_free_space_percent=X.

This does not completely solve the issue but I think this will help user detect this early that thin pool is about to get full better do something about it.

I came to know that lvm tools have similar behavior. They deny creation of new thin volumes if free space in thin pool is below a certain threshold.

**Comment by samuelkarp** [comment]
@rhvgoyal I think that's a good idea.  However, if we're leaving xfs as the default I think it would also be appropriate to add documentation on what happens when your thin pool runs out of space with xfs.

**Comment by rhvgoyal** [comment]
Agreed that documenting xfs behavior on full thin pool is a good idea.

**Comment by samuelkarp** [comment]
@vbatts Can this be reopened to track documentation on the behavior when xfs is used on the thin pool?

**Comment by michipili** [comment]
@snitm What is the resource you like to recommend to users that want to get started with administration of XFS-backed docker deployments?

**Comment by vshefsysorex** [comment]
I know I'm late to this, but I have to ask.  Are thin pools even the right solution for a production environment? How about just a normal LV that is monitored in the same way other partitions are monitored and alerts issued as space runs out.  I also don't see how this is an XFS specific issue.  I get that there is specific behavior that other file systems may not exhibit, but from where I sit, all bets are off when your computer runs out of a finite resource.

Has any one had their OS root partition run out of space?  Bad and unpredictable things happen in that case.  Have you had the partition on which your database is running run out of space?  Ditto.

Why should Docker be expected to behave better than the underlying host OS or other applications?  

I'll also admit that thin pools are a new feature for me, personally, but it seems they just kick the can down the road, and worse, give a false sense of security that disk utilization is "handled" that isn't warranted given that we are still managing a finite resource.

I guess I'm questioning the entire purpose of thin pools, because my concerns should not be restricted to Docker.

**Comment by domq** [comment]
Just a word to the wise here, from a mishap that happened to me yesterday: if you ever find yourself painted in that particular corner, and you are unfortunate enough not to have your `/var/lib/docker` on a separate partition, do *not* reboot hoping that that will reclaim the space. It won't, at least not until you somehow boot up to a rescue image to run `xfs_repair` on your now completely wedged root filesystem ☹

**Comment by rhvgoyal** [comment]
@snitm I had created a PR for this new xfs tunable and it got merged.

https://github.com/docker/docker/pull/26212

**Comment by AaronDMarasco-VSI** [comment]
This hit me again today (on 1.13.0). I am thrilled that it doesn't happen anywhere near as much (thanks to #20786), but it is still a concern with copy-on-write from a container. If this is a fundamental "this is just how XFS works" concern, is there a documented method of recovering?

Right now, the only thing I've been able to do is to give the thin pool a little more disk space (like 100MB) and then reboot. With the new space, `dockerd` at least _starts_, and then I can use `docker rm` / `docker rmi` to clean up.

Is there any other way to recover without nuking all docker storage from orbit and restarting?

**Comment by rhvgoyal** [comment]
power off system, add more space to volume group and then docker should start. Isn't it?

**Comment by AaronDMarasco-VSI** [comment]
@rhvgoyal That works until the VG _really_ runs out of space. That's why I was wondering if there is some other way to examine/clean that thin pool.

**Comment by rhvgoyal** [comment]
thin pool guys have told me that don't let thin pool get full. Otherwise bad things will happen and there are lots of corner cases.

That's the reason we block creation of new containers and pulling of new images if free space left is below threshold. In an attempt to bring this to user's notice.

**Comment by thaJeztah** [comment]
devicemapper was deprecated and removed. 
Let me close this ticket for now, as it looks like it went stale.
